### DISCUSSION TITLE
Create specific mesh from this particular geometry: hollow cylinder divided into radial sectors

### DISCUSSION THREAD

**Discussion by byba1990** [pyvista/pyvista#1986] [Q&A] [answered]
Dear all,

I have the following geometry, read by a file with .vtm extension ( attached here in a zip file):

[Facility.zip](https://github.com/pyvista/pyvista/files/7795823/Facility.zip)

![Schermata da 2021-12-31 11-23-12](https://user-images.githubusercontent.com/45231195/147814951-451f0590-e520-4459-a489-6caddae99feb.png)
![Schermata da 2021-12-31 11-23-26](https://user-images.githubusercontent.com/45231195/147814961-423c705b-68f4-450f-a3b0-b513b33970b2.png)
![Schermata da 2021-12-31 11-23-41](https://user-images.githubusercontent.com/45231195/147814968-4b335add-b863-4131-bb17-9b0faee900c3.png)

It is an holy cylinder, divided into radial sectors. Each sector is separated by a thin empty space. I would like to know, please, if it would be possible to create a mesh, where each section is divided radially (like pieces of a cake)  and along the z-axis. I would provide a representation of the mesh I aim to create. 

![Schermata da 2021-12-31 11-33-00](https://user-images.githubusercontent.com/45231195/147815450-060f0d2a-8950-4895-8d31-0082dcff3b22.png)

In addition, I would like to know how to create the same type of mesh, where the initial geometry is an hollow cylinder, without sectors. The sectors should be created automatically by the division of the mesh This would be a simplified representation of the initial geometry, with the same mesh of the upper figure.

Last thing: I would like to know, please if there is a reference and detailed manual in pdf of pyvista, since I didn't find it online.

Thank you very much in advance for your time.
Best regards,

Christian

**Comment by akaszynski** [comment, marked as answer]
> I would like to know, please, if it would be possible to create a mesh, where each section is divided radially (like pieces of a cake) and along the z-axis

I'd create a single "sector" using a [pyvista.PolyData](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.html) and define several (or a single) line segments via coordinates as specified by the circle equation and then perform an extruded rotation with [extrude_rotate](https://docs.pyvista.org/examples/01-filter/extrude-rotate.html). Then, perform a extrusion in the Z direction with [extrude](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.extrude.html).

> I would like to know, please if there is a reference and detailed manual in pdf of pyvista, since I didn't find it online.

Not yet, but we need to add that.

**Reply by byba1990** [reply]
Dear @akaszynski ,

thank you very much for your reply. I solved the problem by creating the mesh for a single sector in ansys, imported with pyvista, and creating the entire geometry by using `rotate_z` and duplicating the geometry in a loop with `merge`.

Now, I would like to ask for an additional problem that I am facing. The previous geometry needs to assign data to the cells. In particular:

1. I have a particle distribution on the surface of the geometry
2. I would like to assign the particle density (particles/cm^2) to each specific cells and plot the data.

There is a way to associate the particle in that position directly to that specific cell? Since, as I see from [(https://docs.pyvista.org/user-guide/data_model.html)], the assignment of data to cells could be made by a list of data added to `cell_data`. In this method, we do not know the coordinates of the cells and they are enumerated (cells 0, cells 1,...) by default.

Thanks in advance for your time.
Best regards,

Christian

**Reply by MatthewFlamm** [reply]
As I'm understanding, you are asking to find the number of particles, which I assume is from a separate file, in each cell of the existing mesh.  I would use `mesh.find_closest_cell` filter to find the cell closest to the particle position.  Unfortunately, if a particle is outside the mesh and this is possible in your dataset, this filter will still return a cell.  You would have to check for this if you want to avoid counting this particle.

I think PyVista should implement a `mesh.find_containing_cell` filter, which is available through the `FindCell` method of `vtkCellLocator`.  This would only return a cell if it contains the position point.

**Reply by byba1990** [reply]
Ok thank you for your answer. I have written the following function, which finds the cell for that particle position from the bounds i.e.:

```
def getCellIdCoord(position, cellBounds):
    if cellBounds[0]<=position[0]<=cellBounds[1]: 
        if cellBounds[2]<=position[1]<=cellBounds[3]:
            if cellBounds[4]<=position[2]<=cellBounds[5]:
                return True
    else:
                return False
```

After that, I implemented two `for` cycles, the first one over the cells id, the second one over the particle positions and calling the above function. This method works but it takes computational time for many particles.

I used your first suggestion, `mesh.find_closest_cell` but, comparing with the results that I obtained from the function that I created, it misses some particles. After that, I tried to use `mesh.find_containing_cell` and `FindCell` without success. Could you check if there is a way to use it, please? Even though I solved the problem with the function `getCellIdCoord`, I would like to know if there is a method which is faster in term of computational time.
Thank you very much!
Regards,

Christian

**Reply by byba1990** [reply]
Dear @MatthewFlamm and @akaszynski,

I have a problem using mesh.find_closest_cell since this function returns the value -1, even though this particle are inside the cells. Do you know the reason of this behaviour, please?

Thanks in advance for your time.
Best regards,

Christian

**Reply by byba1990** [reply]
PS: there is a way to use the function mesh.find_closest_cell, please? Since I tried it, without sucess.

**Reply by MatthewFlamm** [reply]
What version of pyvista are you using?  Fixes were implemented for `find_closest_cell` in v0.33.0.  I suspect you are using a prior version given the return value of -1.  Although from your description that the particles are always inside a cell (some data may have the particles only just slightly outside the domain), it should have still worked in previous versions.

If using v0.33.0 doesn't work, you'll have to provide a reproducible non working example to investigate.

The documentation has a few examples:
https://docs.pyvista.org/api/core/_autosummary/pyvista.UnstructuredGrid.find_closest_cell.html

**Reply by byba1990** [reply]
Dear @MatthewFlamm,

thank you very much. By updating pyvista, the problem is solved. I should have an additional question: I am working with 1e7 arrays of three entries (position  [x y z]), where `find_closest_cell` is used for each of them in a for cycle to obtain the particle distribution in the mesh.
Here below the for cycle that I implemented to obtain the particle density for the mesh:

```
Dens=np.zeros((nCells))
for pos in posList:
    idCell=target.find_closest_cell(pos)
    boundsCell=target.cell_bounds(idCell)
    Dens[idCell]=Dens[idCell]+1
```

The for cycle requires 45 minutes. Do you have some suggestion, please, to perform this procedure in a faster way?
Thanks in advance for your time!
Best regards,

Christian

**Reply by MatthewFlamm** [reply]
I suspect your bottle neck is in looping with `find_closest_cell`.  

```py
Dens=np.zeros((nCells))
idCells=target.find_closest_cell(posList)
# you probably can vectorize this loop as well if needed
for idCell in idCells:
    Dens[idCell]=Dens[idCell]+1
```

**Reply by byba1990** [reply]
Dear @MatthewFlamm ,

thank you very much. Now, it is definitely faster.
Thank you again for your time.
Best regards,

Christian

**Comment by byba1990** [comment]
Hi all, with respect the probability distribution that I presented, I would like to ask another suggestion, please. In my work, I need to carry out a random generation of particles, considering the probability density distribution, where a single particle has a specific position (x,y,z). Do you have any suggestion, please?
